packages/hardhat-core/src/internal/hardhat-network/provider/modules/eth.ts

```diff
@@ -353,7 +363,11 @@ export class EthModule {
       trace,
       error,
       consoleLogMessages,
-    } = await this._node.runCall(callParams, blockNumberOrPending);
+    } = await this._node.runCall(
+      callParams,
+      blockNumberOrPending,
+      stateOverrideSet
```

Comment:
I would do `stateOverrideSet ?? {}` here, use a mandatory state override object in the rest of the code, and remove the `optionalStateOverrideSet !== undefined` check in `_runTxAndRevertMutations`.